### PR TITLE
Extend IDynamicInterfaceCastable test to cover public default-implemented methods on interface impl types.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Extend IDynamicInterfaceCastable test suite to validate current behavior around calling a default-implemented public method defined on the interface impl type. Today we throw an InvalidCastException because we try to use interface dispatch to resolve the implementation of the method in case the object overrides the implementation.

This behavior was discovered when working on using IDynamicInterfaceCastable in C#/WinRT.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
We should look at documenting the behavior and possibly adding an analyzer to guide people to use `private` interface methods since those work correctly.

File: src/tests/Interop/IDynamicInterfaceCastable/Program.cs [hunk 1]
```diff
@@ -68,6 +69,12 @@ public static int GetNumberStatic()
             return GetNumberStaticReturnValue;
         }
 
+        public static int GetNumberHelperReturnValue = 4;
+        int GetNumberHelper()
+        {
+            return GetNumberHelperReturnValue;
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/IDynamicInterfaceCastable/Program.cs:75]: This appears incorrect. We expect this call to trigger an exception right? This means we should call `Assert.Fail("<DESCRIPTION OF VIOLATION>");`

File: src/tests/Interop/IDynamicInterfaceCastable/Program.cs [hunk 2]
```diff
@@ -81,6 +88,8 @@ int ITest.CallImplemented(ImplementationToCall toCall)
                     return GetNumberPrivate();
                 case ImplementationToCall.InterfaceStatic:
                     return GetNumberStatic();
+                case ImplementationToCall.ImplInterfacePublic:
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/IDynamicInterfaceCastable/Program.cs:91]: > `case ImplementationToCall.ImplInterfacePublic:`

The implementation is `private` right?

Comment (jkoritzinsky) [reply src/tests/Interop/IDynamicInterfaceCastable/Program.cs:91]: By default, interface methods are public. That's why this causes an issue in comparison to the test case with a private method on the impl interface.

Comment (AaronRobinsonMSFT) [reply src/tests/Interop/IDynamicInterfaceCastable/Program.cs:90]: Wait a minute. I thought for default interface implementations by default they were private? I must be misremembering the feature review. Okay, There is no value in relying on the default. Let's be explicit and use the `public` keyword instead.

Comment (jkoritzinsky) [reply src/tests/Interop/IDynamicInterfaceCastable/Program.cs:90]: By default, the default implemented method can only be called on an expression of the interface type. That's what caught us off-guard in C#/WinRT which is why I added the test without the modifier. I'll add the keyword.

File: src/tests/Interop/IDynamicInterfaceCastable/Program.cs [hunk 3]
```diff
@@ -68,6 +69,11 @@ public static int GetNumberStatic()
             return GetNumberStaticReturnValue;
         }
 
+        public int GetNumberHelper()
+        {
+            Asssert.Fail("Calling a public interface method with a default implementation should go through IDynamicInterfaceCastable for interface dispatch.");
+        }
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/IDynamicInterfaceCastable/Program.cs:75]: Doesn't this need a `return` statement?

Comment (AaronRobinsonMSFT) [reply src/tests/Interop/IDynamicInterfaceCastable/Program.cs:75]: Also I think ~~`Asssert`~~ -> `Assert`.

**Comment by jkoritzinsky** [comment]
Any other feedback on this PR?

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**